### PR TITLE
NE-955: IngressController - Add LoadBalancer AllowedSourceRanges API

### PR DESCRIPTION
**Pull request by suleymanakbas91**
- `operator/v1/types_ingress.go` (CIDR): New type. Specify an IP address range in CIDR notation.
(LoadBalancerStrategy): Add AllowedSourceRanges field of type slice of CIDR.
- The rest is generated.

Enhancement Proposal: https://github.com/openshift/enhancements/pull/1177

### CODE REVIEW

File: operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml [hunk 1]
```diff
@@ -217,6 +217,19 @@ spec:
...
+                          ranges to which access to the load balancer should be restricted.  Each
+                          range must be specified using CIDR notation (e.g., "10.0.0.0/8").  If
+                          this field is empty, all source addresses are allowed (which
+                          is equivalent to specifying "0.0.0.0/0").
+                        items:
+                          description: CIDR is an IP address range in CIDR notation
+                            (for example, "10.0.0.0/8" or "fd00::/8").
+                          format: cidr
+                          type: string
+                        nullable: true
```
Comment (jerpeter1) [inline operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml:231]: Consider `nullable: false`, `default: "0.0.0.0/0"`, and removing that part of the description which says that will happen if allowedSourceRanges isn't specified.

File: operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml [hunk 2]
```diff
@@ -217,6 +217,19 @@ spec:
                     description: loadBalancer holds parameters for the load balancer.
                       Present only if type is LoadBalancerService.
                     properties:
+                      allowedSourceRanges:
+                        description: allowedSourceRanges specifies a list of IP address
```
Comment (jerpeter1) [inline operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml:221]: I would use the word "whitelist", "allowlist", or "safelist", somewhere in the description, because those are the words I'd be looking for if I were trying to figure out how to configure this option.

File: operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml [hunk 3]
```diff
@@ -217,6 +217,19 @@ spec:
...
+                          ranges to which access to the load balancer should be restricted.  Each
+                          range must be specified using CIDR notation (e.g., "10.0.0.0/8").  If
+                          this field is empty, all source addresses are allowed (which
+                          is equivalent to specifying "0.0.0.0/0").
+                        items:
+                          description: CIDR is an IP address range in CIDR notation
+                            (for example, "10.0.0.0/8" or "fd00::/8").
+                          format: cidr
```
[the remark this answers is not in the document]
Comment (suleymanakbas91) [reply operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml]: I was not sure about this either. Instead, I found out [this regex](https://github.com/openshift/api/blob/master/network/v1/001-clusternetwork-crd.yaml#L54) being used in multiple places for CIDR, so I re-used it here.

File: operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml (hunk 1 above)
Comment (suleymanakbas91) [reply operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml]: Done.

File: operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml (hunk 2 above)
Comment (suleymanakbas91) [reply operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml]: Good idea! I used allowlist in the description.

**Comment by gcs278** [comment]
No comments, looks good to me, but I'll wait for enhancement proposal to merge before lgtm.

File: operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml (hunk 1 above)
Comment (suleymanakbas91) [reply operator/v1/0000_50_ingress-operator_00-ingresscontroller.crd.yaml:232]: Reverting this in favor of https://github.com/openshift/enhancements/pull/1177#discussion_r930203005.

File: operator/v1/types_ingress.go [hunk 4]
```diff
@@ -367,6 +367,11 @@ var (
 	ExternalLoadBalancer LoadBalancerScope = "External"
 )
 
+// CIDR is an IP address range in CIDR notation (for example, "10.0.0.0/8"
+// or "fd00::/8").
+// +kubebuilder:validation:Pattern=`^(([0-9]|[0-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[0-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])/([0-9]|[12][0-9]|3[0-2])$`
```
Comment (Miciah) [inline operator/v1/types_ingress.go:372]: This pattern doesn't match IPv6 CIDRs.  https://go.dev/play/p/RxELlrqFRT1

File: operator/v1/types_ingress.go [hunk 5]
```diff
@@ -376,6 +381,15 @@ type LoadBalancerStrategy struct {
 	// +required
 	Scope LoadBalancerScope `json:"scope"`
 
+	// allowedSourceRanges specifies an allowlist of IP address ranges to which
+	// access to the load balancer should be restricted.  Each range must be
+	// specified using CIDR notation (e.g., "10.0.0.0/8").  If no range is
+	// specified, "0.0.0.0/0" is used by default, which allows all source addresses.
```
Comment (Miciah) [inline operator/v1/types_ingress.go:387]: IPv6 is allowed by default, right?  So the default is really two entries: "0.0.0.0/0" for IPv4 and "::/0" for IPv6.

Comment (Miciah) [inline operator/v1/types_ingress.go:387]: Document that `service.beta.kubernetes.io/load-balancer-source-ranges` on the service takes precedence.

File: operator/v1/types_ingress.go (hunk 4 above)
Comment (suleymanakbas91) [reply operator/v1/types_ingress.go]: I changed it.

File: operator/v1/types_ingress.go (hunk 5 above)
Comment (suleymanakbas91) [reply operator/v1/types_ingress.go]: Added.

Comment (suleymanakbas91) [reply operator/v1/types_ingress.go]: I changed the comment.

**Comment by suleymanakbas91** [comment]
Playground to test the CIDR range validation pattern: https://go.dev/play/p/kxt-x6LJknW

**Comment by suleymanakbas91** [comment]
Rebased over latest master. Run `make generate-with-container` again and it added more generated files to the PR.

File: operator/v1/types_ingress.go [hunk 6]
```diff
@@ -376,6 +386,19 @@ type LoadBalancerStrategy struct {
...
 	Scope LoadBalancerScope `json:"scope"`
 
+	// allowedSourceRanges specifies an allowlist of IP address ranges to which
+	// access to the load balancer should be restricted.  Each range must be
+	// specified using CIDR notation (e.g. "10.0.0.0/8" or "fd00::/8"). If no range is
+	// specified, "0.0.0.0/0" for IPv4 and "::/0" for IPv6 are used by default,
+	// which allows all source addresses.
+	//
+	// service.beta.kubernetes.io/load-balancer-source-ranges annotation on the
+	// load balancer service takes precedence on OpenShift v4.12.
```
Comment (Miciah) [inline operator/v1/types_ingress.go:396]:
```suggestion
	// To facilitate migration, service.beta.kubernetes.io/load-balancer-source-ranges annotation on the
	// load balancer service takes precedence on OpenShift v4.12 if this field is empty.
```

Comment (suleymanakbas91) [reply operator/v1/types_ingress.go]: Applied.

File: operator/v1/types_ingress.go [hunk 7]
```diff
@@ -367,6 +367,16 @@ var (
...
...rn=`(^(([0-9]|[0-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[0-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])/([0-9]|[12][0-9]|3[0-2])$)|(^s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]d|1dd|[1-9]?d)(.(25[0-5]|2[0-4]d|1dd|[1-9]?d)){3}))|:)))(%.+)?s*(\/(12[0-8]|1[0-1][0-9]|[1-9][0-9]|[0-9]))$)`
```
Comment (deads2k) [inline operator/v1/types_ingress.go:372]: :exploding_head:

File: operator/v1/types_ingress.go [hunk 8]
```diff
@@ -376,6 +386,21 @@ type LoadBalancerStrategy struct {
...
 
+	// allowedSourceRanges specifies an allowlist of IP address ranges to which
+	// access to the load balancer should be restricted.  Each range must be
+	// specified using CIDR notation (e.g. "10.0.0.0/8" or "fd00::/8"). If no range is
+	// specified, "0.0.0.0/0" for IPv4 and "::/0" for IPv6 are used by default,
+	// which allows all source addresses.
+	//
+	// To facilitate migration from earlier versions of OpenShift that did not have the
+	// allowedSourceRanges field, the service.beta.kubernetes.io/load-balancer-source-ranges
+	// annotation on the load balancer service takes precedence on OpenShift v4.12
```
Comment (deads2k) [inline operator/v1/types_ingress.go:397]: which service is the load balancer service?

Comment (deads2k) [inline operator/v1/types_ingress.go:397]: it takes precedence, it's not used in combination, so if I specify value/A here and value/B on the annotation, I get value/B and this gets ignored?

Can we write a control loop that actually stomps this when that is happening so that a user knows the state of the cluster?  Otherwise a client cannot rely on this API and must interpret two different APIs to determine his configuration.

Comment (Miciah) [reply operator/v1/types_ingress.go:397]: To David's point, we can clarify as follows:
```suggestion
 	// To facilitate migration from earlier versions of OpenShift that did
	// not have the allowedSourceRanges field, you may set the
	// service.beta.kubernetes.io/load-balancer-source-ranges annotation on
	// the "router-<ingresscontroller name>" service in the
	// "openshift-ingress" namespace, and this annotation will take
	// precedence over allowedSourceRanges on OpenShift 4.12.
```

Comment (Miciah) [reply operator/v1/types_ingress.go:397]: > Can we write a control loop that actually stomps this when that is happening so that a user knows the state of the cluster? Otherwise a client cannot rely on this API and must interpret two different APIs to determine his configuration.

`allowedSourceRanges` appears in `spec` and in `status`.  Would it be appropriate to reflect the annotation (if it's set) to `status.allowedSourceRanges` when `spec.allowedSourceRanges` is empty?

Comment (suleymanakbas91) [reply operator/v1/types_ingress.go]: I applied Miciah's suggestion.

Comment (suleymanakbas91) [reply operator/v1/types_ingress.go]: I've updated the [EP](https://github.com/openshift/enhancements/pull/1177/files#diff-2182da078aabca9029bf489d67b4e726c7fc72fcf4878cae69a6a1d08f3ff249R138) to state `status.allowedSourceRanges` will reflect the current effective value by looking at the annotation and the `spec.loadBalancerSourceRanges` field.

**Comment by deads2k** [comment]
if you want to test your regex, @JoelSpeed created a way to provide a test manifest and expected result files.

**Comment by JoelSpeed** [comment]
> if you want to test your regex, @JoelSpeed created a way to provide a test manifest and expected result files.

Some examples of the most minimal tests can be found in #1305, from there we can add additional test cases to exercise various expected positive and negative validations, lmk if you want to try this out
